pg_search/src/postgres/customscan/hook.rs

```diff
@@ -581,6 +589,69 @@ unsafe fn query_has_search_operator(parse: *mut pg_sys::Query) -> bool {
     false
 }
 
+/// Recursively check if an expression tree contains any ParadeDB search operator.
+/// Uses expression tree walker to examine all OpExpr nodes.
+unsafe fn expr_contains_paradedb_operator(node: *mut pg_sys::Node) -> bool {
+    struct WalkerContext {
+        found: bool,
+    }
+
+    #[pg_guard]
+    unsafe extern "C-unwind" fn walker(
+        node: *mut pg_sys::Node,
+        context: *mut core::ffi::c_void,
+    ) -> bool {
+        if node.is_null() {
+            return false;
+        }
+
+        let ctx = context.cast::<WalkerContext>();
+
+        // Check if this is an OpExpr
+        if let Some(opexpr) = nodecast!(OpExpr, T_OpExpr, node) {
+            if is_paradedb_search_operator((*opexpr).opno) {
+                (*ctx).found = true;
+                return true; // Stop walking
+            }
+        }
+
+        // Continue walking the tree
+        pg_sys::expression_tree_walker(node, Some(walker), context)
+    }
+
+    let mut context = WalkerContext { found: false };
+    walker(node, &mut context as *mut _ as *mut core::ffi::c_void);
+    context.found
+}
+
+/// Check if an operator OID is a ParadeDB search operator.
+///
+/// Checks operator name regardless of argument types (text, text[], pdb.query, pdb.boost, pdb.fuzzy, etc.)
+unsafe fn is_paradedb_search_operator(opno: pg_sys::Oid) -> bool {
```

Comment:
Is this worth memoizing? We're not really optimizing per-query overhead right now, so I can't really tell whether we'd see it in a profile. So probably the answer is: "let's not worry about it yet".